cpp/modmesh/toggle/callprofiler.hpp

```diff
@@ -0,0 +1,212 @@
+    {
+        auto start_time = std::chrono::high_resolution_clock::now();
+        m_radix_tree.entry(caller_name);
+        CallerProfile & callProfile = m_radix_tree.get_current_node()->data();
+        callProfile.caller_name = caller_name;
+        callProfile.start_stopwatch();
+    }
+
+    // Called when a function ends
+    void end_caller(const std::string & caller_name)
+    {
+        CallerProfile & callProfile = m_radix_tree.get_current_node()->data();
+
+        // Update profiling information
+        callProfile.stop_stopwatch();
+
+        // Pop the caller from the call stack
+        m_radix_tree.move_current_to_parent();
+    }
+
+    /// Print the profiling information
+    void print_profiling_result(std::ostream & outstream) const
+    {
+        _print_profiling_result(*(m_radix_tree.get_current_node()), 0, outstream);
+    }
+
+    /// Result the profiler
+    void reset()
+    {
+        RadixTreeNode<CallerProfile> * newNode;
+        RadixTreeNode<CallerProfile> * currentNode = m_radix_tree.get_current_node();
+
+        while (!m_radix_tree.is_root())
+        {
+            CallerProfile & profile = m_radix_tree.get_current_node()->data();
+            if (profile.is_running)
+            {
+                profile.stop_stopwatch();
+            }
+            m_radix_tree.move_current_to_parent();
+        }
+        m_radix_tree.reset();
+    }
+
+private:
+    CallProfiler() = default;
+
+    void _print_profiling_result(const RadixTreeNode<CallerProfile> & node, const int depth, std::ostream & outstream) const
+    {
+        for (int i = 0; i < depth; ++i)
+        {
+            outstream << "  ";
+        }
+
+        auto profile = node.data();
+
+        if (depth == 0)
+        {
+            outstream << "Profiling Result" << std::endl;
+        }
+        else
+        {
+            outstream << profile.caller_name << " - Total Time: " << profile.total_time.count() / 1e6 << " ms, Call Count: " << profile.call_count << std::endl;
+        }
+
+        for (const auto & child : node.children())
+        {
+            _print_profiling_result(*child, depth + 1, outstream);
+        }
+    }
+
+private:
+    RadixTree<CallerProfile> m_radix_tree; /// the data structure of the callers
+
+    friend CallProfilerTest;
+};
+
+/// Utility to profile a call
+class CallProfilerProbe
+{
+public:
+    CallProfilerProbe(CallProfiler & profiler, const char * caller_name)
+        : m_profiler(profiler)
+        , m_caller_name(caller_name)
+    {
+        auto cancel_callback = [&]()
+        {
+            cancel();
+        };
+        m_profiler.start_caller(m_caller_name, cancel_callback);
+    }
+
+    ~CallProfilerProbe()
+    {
+        if (!m_cancel)
+        {
+            m_profiler.end_caller(m_caller_name);
+        }
+    }
+
+    void cancel()
+    {
+        m_cancel = true;
+    }
+
+private:
+    const char * m_caller_name;
+    bool m_cancel = false;
+    CallProfiler & m_profiler;
```

Comment:
Because of this reference member `m_profiler` and the const member `m_caller_name`, you should delete implicitly created constructors.

Reply:
done